### PR TITLE
Adding AuthGemJoin

### PR DESCRIPTION
**Pull request by hexonaut**
Doing some cleanup/modernizing along with adding the regular gem join authed version.

### CODE REVIEW

File: src/join-5-auth.sol [hunk 1]
```diff
@@ -34,49 +32,58 @@ interface GemLike {
...
+        uint256 dec_ = dec = GemLike(gem_).decimals();
+        require(dec_ < 18, "GemJoin5/decimals-18-or-higher");
         wards[msg.sender] = 1;
+        emit Rely(msg.sender);
         live = 1;
         vat = VatLike(vat_);
         ilk = ilk_;
     }
 
-    function cage() external note auth {
```
Comment (brianmcmichael) [inline src/join-5-auth.sol:60]: we need an event emission here if we're removing note

File: src/psm.t.sol [hunk 2]
```diff
@@ -35,7 +35,7 @@ interface Hevm {
 
 contract TestToken is DSToken {
 
-    constructor(bytes32 symbol_, uint256 decimals_) public DSToken(symbol_) {
+    constructor(string memory symbol_, uint8 decimals_) public DSToken(symbol_) {
```
Comment (brianmcmichael) [inline src/psm.t.sol:38]: generally I'd prefer to keep the input param a uint256 and then check that it's LT uint8 max

**Comment by brianmcmichael** [comment]
Not sure how I feel about updating code for a join that already has deployed instances.

Users and tests needing to reference old versions are now different from what's on mainnet, plus maintenance burden of having multiple instances of the same nominal adapter across types. Event processing could be problematic if the results differ between contracts.

Would almost rather have an entirely new adapter added.

File: src/join-5-auth.sol (hunk 1 above)
Comment (hexonaut) [reply src/join-5-auth.sol:60]: This should be added to here as well then: https://github.com/makerdao/dss-gem-joins/blob/master/src/join-managed.sol

I used that as an example.

File: src/psm.t.sol (hunk 2 above)
Comment (hexonaut) [reply src/psm.t.sol:38]: Does it really matter for the test with a hardcoded value though?

**Comment by hexonaut** [comment]
Agree having a "psm join adapter" all in one makes more sense, but I figured we'd stick with this for consistency. The join adapter doesn't quite match the existing ones in `dss-gem-joins` which is why I opted to include them in here.

File: src/join-5-auth.sol (hunk 1 above)
Comment (hexonaut) [reply src/join-5-auth.sol:60]: Done.

Comment (brianmcmichael) [reply src/join-5-auth.sol:60]: New issue https://github.com/makerdao/dss-gem-joins/issues/23

File: src/psm.t.sol (hunk 2 above)
Comment (brianmcmichael) [reply src/psm.t.sol:38]: I suppose it's fine in this case. There's not going to be much chance this is going to be deployed manually. As a general rule we probably want to stick to 32-byte data params when possible.

File: src/join-auth.sol [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        gem = GemLike(gem_);
+        dec = GemLike(gem_).decimals();
+    }
+    function cage() external auth {
+        live = 0;
+        emit Cage();
+    }
+    function join(address urn, uint wad, address msgSender) external auth {
+        require(live == 1, "GemJoin/not-live");
+        require(int(wad) >= 0, "GemJoin/overflow");
```
Comment (talbaneth) [inline src/join-auth.sol:69]: Checking `wad <= (2 ** 255 - 1)` is better matching the overflow check on exit().
This is a comment I got in join-managed review:
https://github.com/makerdao/dss-gem-joins/pull/21

File: src/join-auth.sol [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    event Exit(address indexed guy, uint256 wad);
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        gem = GemLike(gem_);
+        dec = GemLike(gem_).decimals();
```
Comment (talbaneth) [inline src/join-auth.sol:61]: dec is unused

File: src/join-auth.sol [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+interface GemLike {
+    function decimals() external view returns (uint);
+    function transfer(address,uint) external returns (bool);
+    function transferFrom(address,address,uint) external returns (bool);
+}
+
+interface VatLike {
+    function slip(bytes32,address,int) external;
+    function move(address,address,uint) external;
```
Comment (talbaneth) [inline src/join-auth.sol:28]: move is unused

File: src/join-auth.sol [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    bytes32 public immutable ilk;   // Collateral Type
+    GemLike public immutable gem;
+    uint    public immutable dec;
+    uint    public live;  // Active Flag
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Cage();
+    event Join(address indexed urn, uint256 wad, address indexed msgSender);
```
Comment (talbaneth) [inline src/join-auth.sol:51]: I would prefer indexed parameters to come first in the event

**Review by talbaneth — COMMENTED** [review]
See minor comments.

File: src/join-auth.sol (hunk 4 above)
Comment (brianmcmichael) [reply src/join-auth.sol:61]: we should keep this for compatibility across adapters

File: src/join-auth.sol [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        require(gem.transferFrom(msgSender, address(this), wad), "GemJoin/failed-transfer");
+        emit Join(urn, wad, msgSender);
+    }
+    function exit(address guy, uint wad) external {
+        require(wad <= 2 ** 255, "GemJoin/overflow");
+        vat.slip(ilk, msg.sender, -int(wad));
+        require(gem.transfer(guy, wad), "GemJoin/failed-transfer");
+        emit Exit(guy, wad);
+    }
+}
```
Comment (brianmcmichael) [inline src/join-auth.sol]: nit: missing newline

File: src/join-auth.sol (hunk 4 above)
Comment (brianmcmichael) [reply src/join-auth.sol:61]: indeed it's used here: https://github.com/BellwoodStudios/dss-psm/blob/master/src/psm.sol#L72

**Comment by talbaneth** [comment]
Approved for merge (the current version is what got deployed eventually, and suggested changes are minor).

**Pull request merged**